### PR TITLE
Adds createRef and forwardRef to the React compiler

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: Adds`createRef` and `forwardRef` to the React compiler

This PR adds both features from React 16.3, including full inlining of forwardRef functions.

### CODE REVIEW

File: test/react/functional-components/refs.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  var x = React.createRef();
+
+  return (
+    <WrappedComponent ref={x} />
+  );
+}
+
+App.getTrials = function(renderer, Root) {
+  renderer.update(<Root />);
+  return [['simple render with refs', renderer.toJSON()]];
```
Comment (gaearon) [inline test/react/functional-components/refs.js]: It's not obvious from this test that the ref forwarding actually works. Can you make it do something with that ref? e.g. pass it down a few levels to a class, call a method to read a prop.

Comment (trueadm) [reply test/react/functional-components/refs.js]: I originally tried, but the `test-renderer` doesn't seem to work with object refs right now so I was going to look into that after this landed.

Comment (trueadm) [reply test/react/functional-components/refs.js]: I'll add a callback ref to show it working though.

File: src/react/utils.js [hunk 2]
```diff
@@ -649,11 +652,13 @@ export function getProperty(
   property: string | SymbolValue
 ): Value {
   if (object instanceof AbstractObjectValue) {
-    let elements = object.values.getElements();
+    let elements = object.values._elements;
```
Comment (gaearon) [inline src/react/utils.js]: Is this a private field?

File: test/react/functional-components/refs2.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    <WrappedComponent ref={props.x} />
+  );
+}
+
+App.getTrials = function(renderer, Root) {
+  let results = [];
+  var refValue = null;
+  renderer.update(<Root x={val => refValue = val} />);
+  results.push(['simple render with refs', renderer.toJSON()]);
+  results.push(['ref node is of div', refValue.type]);
```
Comment (gaearon) [inline test/react/functional-components/refs2.js]: Can we also verify that:

* You can forward refs more through several levels
* It works with class instances too (e.g. test `ref.current.getValue()` with `getValue() { return this.props.value; }`)

File: src/react/utils.js (hunk 2 above)
Comment (trueadm) [reply src/react/utils.js]: Yep, but `getElements` triggers an invariant if it's undefined, which is why we can't use it here.

Comment (gaearon) [reply src/react/utils.js]: Is there a good reason why it triggers an invariant? I suppose somebody put it there to protect against something. Why should we get around that?

File: test/react/functional-components/refs2.js (hunk 3 above)
Comment (trueadm) [reply test/react/functional-components/refs2.js]: - You can forward refs more through several levels

Can you give me an example of a test like this? Ideally, if you could just add a said test to this PR that would be awesome.

Comment (gaearon) [reply test/react/functional-components/refs2.js]: `forwardRef` that renders another `forwardRef`

Comment (gaearon) [reply test/react/functional-components/refs2.js]: Maybe this doesn't make sense. I'm not sure.

File: src/react/utils.js (hunk 2 above)
Comment (trueadm) [reply src/react/utils.js]: Fixed. :)

File: src/evaluators/JSXElement.js [hunk 4]
```diff
@@ -308,6 +316,13 @@ function evaluateJSXAttributes(
       );
       if (!mayContainRefOrKey) {
         deleteRefAndKeyFromProps(realm, config);
+      } else {
+        if (keyValue !== undefined) {
+          setConfigProperty("key", keyValue);
+        }
+        if (refValue !== undefined) {
+          setConfigProperty("ref", refValue);
+        }
```
[the remark this answers is not in the document]
Comment (trueadm) [reply src/evaluators/JSXElement.js]: Found a bug when you pass in `key` or `ref` to a ReactElement that is partial but we know the `key` or `ref` to be concrete, i.e. `<div {...partialObject} key="123" />`

**Pull request closed without merging**